elfpy/data/contract_interface.py

```diff
@@ -126,46 +186,84 @@ def get_event_object(
-    # get the callable contract function from function_name & call it
-    function: ContractFunction = contract.get_function_by_name(function_name)()
-    return_values = function.call(**function_args)
-    # associate returned values with the keys
-    assert len(return_value_keys) == len(return_values)
-    function_return_dict = dict((variable_name, info) for variable_name, info in zip(return_value_keys, return_values))
-    return function_return_dict
-
-
-def initialize_web3_with_http_provider(ethereum_node: URI | str, request_kwargs: dict | None = None) -> Web3:
-    """Initialize a Web3 instance using an HTTP provider and inject a geth Proof of Authority (poa) middleware.
-
-    Arguments
-    ---------
-    ethereum_node: URI | str
-        Address of the http provider
-    request_kwargs: dict
-        The HTTPProvider uses the python requests library for making requests.
-        If you would like to modify how requests are made,
-        you can use the request_kwargs to do so.
+def contract_function_abi_outputs(contract_abi: ABI, function_name: str) -> list[tuple[str, str]] | None:
+    """Parse the function abi to get the name and type for each output"""
+    function_abi = None
+    for abi in contract_abi:  # loop over each entery in the abi list
+        if abi.get("name") == function_name:  # check the name
+            function_abi = abi  # pull out the one with the desired name
+            break
+    if function_abi is None:
+        logging.warning("could not find function_name=%s in contract abi", function_name)
+        return None
+    function_outputs = function_abi.get("outputs")
+    if function_outputs is None:
+        logging.warning("function abi does not specify outputs")
+        return None
+    if not isinstance(function_outputs, Sequence):
+        logging.warning("function abi outputs are not a sequence")
+        return None
+    if len(function_outputs) > 1:  # multiple unnamed vars were returned
+        return_names_and_types = []
+        for output in function_outputs:
+            return_names_and_types.append(_get_name_and_type_from_abi(output))
+    if (
+        function_outputs[0].get("type") == "tuple" and function_outputs[0].get("components") is not None
+    ):  # multiple named outputs were returned in a struct
+        abi_components = function_outputs[0].get("components")
+        if abi_components is None:
+            logging.warning("function abi output componenets are not a included")
+            return None
+        return_names_and_types = []
+        for component in abi_components:
+            return_names_and_types.append(_get_name_and_type_from_abi(component))
+    else:  # final condition is a single output
+        return_names_and_types = [_get_name_and_type_from_abi(function_outputs[0])]
+    return return_names_and_types
+
+
+def smart_contract_read(contract: Contract, function_name: str, *fn_args, **fn_kwargs) -> dict[str, Any]:
```

Comment:
nice, lets add a todo for some tests here too.  wanna test the return types specifically.